libzecale/circuits/aggregator_gadget.tcc

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef __ZECALE_CIRCUITS_AGGREGATOR_GADGET_TCC__
+#define __ZECALE_CIRCUITS_AGGREGATOR_GADGET_TCC__
+
+#include "libzecale/circuits/aggregator_gadget.hpp"
+
+namespace libzecale
+{
+
+template<
+    typename nppT,
+    typename wppT,
+    typename nsnarkT,
+    typename wverifierT,
+    size_t NumProofs>
+aggregator_gadget<nppT, wppT, nsnarkT, wverifierT, NumProofs>::
+    aggregator_gadget(
+        libsnark::protoboard<libff::Fr<wppT>> &pb,
+        const size_t inputs_per_nested_proof,
+        const std::string &annotation_prefix)
+    : libsnark::gadget<libff::Fr<wppT>>(pb, annotation_prefix)
+    , num_inputs_per_nested_proof(inputs_per_nested_proof)
+{
+    // Assert that a single nested input can be encoded in a single wrapper
+    // input. This holds for all pairing chains we test, but if it fails then
+    // nested inputs will need to be encoded over multiple wrapper inputs.
+    // Unfortunately this check cannot be made statically.
+    //
+    // TODO: if num_bits are equal then also check mod values.
```

Comment:
What do you mean by "mod values"? If you want to check that the nested inputs are indeed elements of $\FF_{r_{BLS12377}}$, then doing a range check here (in the statement) will be quite expensive. This check is carried out on the smart contract.

Reply:
Here `mod` refers to the static member of the fields.  This is concerned with the assert on the following like, which currently only checks the `num_bits` member.  It's not related to checks on the contract or in the circuit, but an up-front check that the encoding from Fr<nppT> into Fr<wppT> is injective (as described above).

Reply:
I've reworded to try and make this clearer.

Reply:
Got it, I see.